indra/llwindow/llgamecontrol.cpp

```diff
@@ -515,51 +616,104 @@ void LLGameControllerManager::removeController(SDL_JoystickID id)
-    if (it != mDevices.end())
+    // Map axis using device-specific settings
+    // or leave the value unchanged
+    U8 mapped_axis = it->mOptions.mapAxis(axis);
+    if (mapped_axis != axis)
     {
-        // Note: the RAW analog joysticks provide NEGATIVE X,Y values for LEFT,FORWARD
-        // whereas those directions are actually POSITIVE in SL's local right-handed
-        // reference frame.  Therefore we implicitly negate those axes here where
-        // they are extracted from SDL, before being used anywhere.
-        if (axis < SDL_CONTROLLER_AXIS_TRIGGERLEFT)
-        {
-            // Note: S16 value is in range [-32768, 32767] which means
-            // the negative range has an extra possible value.  We need
-            // to add (or subtract) one during negation.
-            if (value < 0)
-            {
-                value = - (value + 1);
-            }
-            else if (value > 0)
-            {
-                value = (-value) - 1;
-            }
-        }
+        LL_DEBUGS("SDL2") << "Axis mapped: joystick=0x" << std::hex << id << std::dec
+            << " input axis i=" << (S32)axis
+            << " mapped axis i=" << (S32)mapped_axis << LL_ENDL;
+        axis = mapped_axis;
+    }
 
-        LL_DEBUGS("SDL2") << "joystick=0x" << std::hex << id << std::dec
+    if (axis > MAX_AXIS)
+    {
+        LL_WARNS("SDL2") << "Unknown axis: joystick=0x" << std::hex << id << std::dec
             << " axis=" << (S32)(axis)
             << " value=" << (S32)(value) << LL_ENDL;
-        it->mState.mAxes[axis] = value;
+        return;
+    }
+
+    // Fix value using device-specific settings
+    // or leave the value unchanged
+    S16 fixed_value = it->mOptions.fixAxisValue(axis, value);
+    if (fixed_value != value)
+    {
+        LL_DEBUGS("SDL2") << "Value fixed: joystick=0x" << std::hex << id << std::dec
+            << " axis i=" << (S32)axis
+            << " input value=" << (S32)value
+            << " fixed value=" << (S32)fixed_value << LL_ENDL;
+        value = fixed_value;
     }
+
+    // Note: the RAW analog joysticks provide NEGATIVE X,Y values for LEFT,FORWARD
+    // whereas those directions are actually POSITIVE in SL's local right-handed
+    // reference frame.  Therefore we implicitly negate those axes here where
+    // they are extracted from SDL, before being used anywhere.
+    if (axis < SDL_CONTROLLER_AXIS_TRIGGERLEFT)
+    {
+        // Note: S16 value is in range [-32768, 32767] which means
+        // the negative range has an extra possible value.  We need
+        // to add (or subtract) one during negation.
+        if (value < 0)
+        {
+            value = - (value + 1);
+        }
+        else if (value > 0)
+        {
+            value = (-value) - 1;
+        }
+    }
+
+    LL_DEBUGS("SDL2") << "joystick=0x" << std::hex << id << std::dec
+        << " axis=" << (S32)(axis)
+        << " value=" << (S32)(value) << LL_ENDL;
+    it->mState.mAxes[axis] = value;
```

Comment:
We might want to store the axis data in its raw un-mapped order and then perform the re-mapping when it comes time to actually use it (e.g when it comes time to put the value on the wire to the server, or to apply it to the avatar).  However, for now this is fine: it is good to get the options parsing done and to stub the places where we'll use it.  Full speed ahead.